app/modules/gpio_pulse.c

```diff
@@ -137,6 +137,69 @@ static int gpio_pulse_delete(lua_State *L) {
   return 0;
 }
 
+static void fill_entry_from_table(lua_State *L, pulse_entry_t *entry) {
+  if (lua_type(L, -1) != LUA_TTABLE) {
+    luaL_error(L, "All entries must be tables");
+  }
+
+  lua_pushnil(L);             // key position
+  while (lua_next(L, -2)) {
+    // stack now contains: -1 => value; -2 => key; -3 => table
+    if (lua_type(L, -2) == LUA_TNUMBER) {
+      int pin = lua_tonumber(L, -2);
+      int value = lua_tonumber(L, -1);
+
+      if (pin < 0 || pin >= GPIO_PIN_NUM) {
+        luaL_error(L, "pin number %d must be in range 0 .. %d", pin, GPIO_PIN_NUM - 1);
+      }
+
+      if (value) {
+        entry->gpio_set |= BIT(pin_num[pin]);
+      } else {
+        entry->gpio_clr |= BIT(pin_num[pin]);
+      }
+    } else {
+      const char *str = lua_tostring(L, -2);
+
+      if (strcmp(str, "delay") == 0) {
+        entry->delay = lua_tonumber(L, -1);
+        if (entry->delay < 0 || entry->delay > DELAY_LIMIT) {
```

Comment:
I expect that this won't throw an error if value at index `-1` is not convertible to a number. [`lua_tonumber()`](http://pgl.yoyo.org/luai/i/lua_tonumber) returns 0 in this case and the range check is happy.

`luaL_checknumber()` or an explicit test for `lua_isnumber()` if delay == 0 would detect nonsense.

Applies to further instances of `lua_tonumber()` below.